File: backend-express/src/services/menuItemService.ts
import { Prisma } from "@prisma/client";
import type { MenuItemListSortedByFields } from "@shared/dtos";
import type { MenuItemListRequestDto, MenuItemUpsertRequestDto } from "@shared/dtos";
import type { MenuItemListResponseDto, MenuItemDetailResponseDto } from "@shared/dtos";
import { responseDtoFactory } from "@/dtos";
import { NotFoundError, OperationError } from "@shared/errors";
import { BaseService } from "./baseService";
import type { IMenuItemService } from "@shared/services/menuItemService";

export class MenuItemService extends BaseService implements IMenuItemService {
  public async getListAsync(
    requestDto?: MenuItemListRequestDto): Promise<MenuItemListResponseDto> {
    // Default request values.
    const {
      sortedByAscending = true,
      sortedByField = "name",
      page = 1,
      resultsPerPage = 15
    } = requestDto ?? { };

    // Compute result count.
    const resultCount = await this.prisma.menuItem.count();
    if (!resultCount) {
      return responseDtoFactory.menuItem.createList(0, []);
    }

    // Compute field and direction to sort.
    const sortedByConditions: SortedByConditions<MenuItemListSortedByFields> = { };
    switch (sortedByField) {
      case "name":
        sortedByConditions.name = sortedByAscending ? "asc" : "desc";
        break;
      case "defaultAmount":
        sortedByConditions.defaultAmount = sortedByAscending ? "asc" : "desc";
        break;
      case "defaultVatPercentage":
        sortedByConditions.defaultVatPercentage = sortedByAscending ? "asc" : "desc";
        break;
      case "createdDateTime":
        sortedByConditions.createdDateTime = sortedByAscending ? "asc" : "desc";
        break;
    }

    // Fetch the entities from the database.
    const menuItems = await this.prisma.menuItem.findMany({
      include: { category: true },
      orderBy: sortedByConditions,
      skip: resultsPerPage * (page - 1),
      take: resultsPerPage
    });

    const pageCount = Math.ceil(resultCount / resultsPerPage);
    return responseDtoFactory.menuItem.createList(pageCount, menuItems);
  }

  public async getDetailAsync(id: number): Promise<MenuItemDetailResponseDto> {
    const menuItem = await this.prisma.menuItem.findFirst({
      where: { id },
      include: { category: true }
    });

    if (!menuItem) {
      throw new NotFoundError();
    }

    return responseDtoFactory.menuItem.createDetail(menuItem);
  }

  public async createAsync(requestDto: MenuItemUpsertRequestDto): Promise<number> {
    try {
      const menuItem = await this.prisma.menuItem.create({
        data: {
          name: requestDto.name,
          defaultAmount: requestDto.defaultAmount,
          defaultVatPercentage: requestDto.defaultVatPercentage,
          unit: requestDto.unit,
          thumbnailUrl: requestDto.thumbnailUrl,
          createdDateTime: new Date(),
          categoryId: requestDto.categoryId
        }
      });

      return menuItem.id;
    } catch (error) {
      if (error instanceof Prisma.PrismaClientKnownRequestError) {
        const handledResult = this.prismaErrorHandler.handle(error);
        if (handledResult.isForeignKeyConstraintError) {
          throw OperationError.notFound("categoryId");
        }
      }

      throw error;
    }
  }

  public async updateAsync(id: number, requestDto: MenuItemUpsertRequestDto): Promise<void> {
    try {
      await this.prisma.menuItem.update({
        where: { id },
        data: {
          name: requestDto.name,
          defaultAmount: requestDto.defaultAmount,
          defaultVatPercentage: requestDto.defaultVatPercentage,
          unit: requestDto.unit,
          thumbnailUrl: requestDto.thumbnailUrl,
          lastUpdatedDateTime: new Date(),
          categoryId: requestDto.categoryId
        }
      });
    } catch (error) {
      if (error instanceof Prisma.PrismaClientKnownRequestError) {
        const handledResult = this.prismaErrorHandler.handle(error);
        if (handledResult.isNotFoundError) {
          throw new NotFoundError();
        }

        if (handledResult.isForeignKeyConstraintError) {
          throw OperationError.notFound("categoryId");
        }
      }

      throw error;
    }
  }

  public async deleteAsync(id: number): Promise<void> {
    try {
      await this.prisma.menuItem.delete({ where: { id } });
    } catch (error) {
      if (error instanceof Prisma.PrismaClientKnownRequestError) {
        const handledResult = this.prismaErrorHandler.handle(error);
        if (handledResult.isNotFoundError) {
          throw new NotFoundError();
        }
      }

      throw error;
    }
  }
}

File: backend-dotnet/Services/MenuItemService.cs
using Microsoft.EntityFrameworkCore;
using MySqlConnector;
using QuanNhauSanVuon.Dtos;
using QuanNhauSanVuon.Extensions;
using QuanNhauSanVuon.Localization;
using QuanNhauSanVuon.Services.Entities;
using QuanNhauSanVuon.Services.Enums;
using QuanNhauSanVuon.Services.Exceptions;
using QuanNhauSanVuon.Services.Interfaces;

namespace QuanNhauSanVuon.Services;

/// <inheritdoc />
public class MenuItemService : IMenuItemService
{
    private readonly DatabaseContext _context;
    private readonly MySqlExceptionHandler _exceptionHandler;

    public MenuItemService(
            DatabaseContext context,
            MySqlExceptionHandler exceptionHandler)
    {
        _context = context;
        _exceptionHandler = exceptionHandler;
    }

    /// <inheritdoc />
    public async Task<MenuItemListDto> GetListAsync(MenuItemListFiltersDto filtersDto)
    {
        int resultCount = await _context.MenuItems.CountAsync();
        if (resultCount == 0)
        {
            return new MenuItemListDto();
        }

        // Compute field and direction to sort.
        IQueryable<MenuItem> query = _context.MenuItems;
        switch (filtersDto.SortedByField)
        {
            case nameof(SortedByFieldOption.Name):
                query = filtersDto.SortedByAscending
                    ? query.OrderBy(mi => mi.Name).ThenBy(mi => mi.CreatedDateTime)
                    : query.OrderByDescending(mi => mi.Name).ThenBy(mi => mi.CreatedDateTime);
                break;
            case nameof (SortedByFieldOption.DefaultNetPrice):
                query = filtersDto.SortedByAscending
                    ? query.OrderBy(mi => mi.DefaultNetPrice).ThenBy(mi => mi.Name)
                    : query.OrderByDescending(mi => mi.DefaultNetPrice).ThenBy(mi => mi.Name);
                break;
            case nameof (SortedByFieldOption.DefaultVatPercentage):
                query = filtersDto.SortedByAscending
                    ? query.OrderBy(mi => mi.DefaultVatPercentage)
                        .ThenBy(mi => mi.CreatedDateTime)
                    : query.OrderByDescending(mi => mi.DefaultVatPercentage)
                        .ThenBy(mi => mi.CreatedDateTime);
                break;
            case nameof (SortedByFieldOption.CreatedDateTime):
                query = filtersDto.SortedByAscending
                    ? query.OrderBy(mi => mi.CreatedDateTime).ThenBy(mi => mi.Name)
                    : query.OrderByDescending(mi => mi.CreatedDateTime).ThenBy(mi => mi.Name);
                break;
            default:
                throw new NotImplementedException();
        }

        // Compute category to filter.
        if (filtersDto.CategoryId.HasValue)
        {
            query = query.Where(mi => mi.CategoryId == filtersDto.CategoryId.Value);
        }

        int pageCount = (int)Math.Ceiling((double)resultCount / filtersDto.ResultsPerPage);
        List<MenuItem> menuItems = await query
            .Skip(filtersDto.ResultsPerPage * (filtersDto.Page - 1))
            .Take(filtersDto.ResultsPerPage)
            .AsSplitQuery()
            .ToListAsync();

        return new MenuItemListDto(pageCount, menuItems);
    }

    /// <inheritdoc />
    public async Task<MenuItemDetailDto> GetDetailAsync(int id)
    {
        return await _context.MenuItems
            .Where(mi => mi.Id == id)
            .Select(mi => new MenuItemDetailDto(mi))
            .SingleOrDefaultAsync()
            ?? throw new NotFoundException();
    }

    /// <inheritdoc />
    public async Task<int> CreateAsync(MenuItemUpsertDto upsertDto)
    {
        try
        {
            MenuItem menuItem = new MenuItem
            {
                Name = upsertDto.Name,
                DefaultNetPrice = upsertDto.DefaultNetPrice,
                DefaultVatPercentage = upsertDto.DefaultVatPercentage,
                Unit = upsertDto.Unit,
                ThumbnailUrl = upsertDto.ThumbnailUrl,
                CategoryId = upsertDto.CategoryId
            };

            _context.MenuItems.Add(menuItem);
            await _context.SaveChangesAsync();
            return menuItem.Id;
        }
        catch (DbUpdateException exception)
        {
            if (exception.InnerException is not MySqlException mySqlException)
            {
                throw;
            }

            bool isConvertible = TryConvertingDbException(
                mySqlException,
                out OperationException convertedException);
            if (!isConvertible)
            {
                throw;
            }

            throw convertedException;
        }
    }

    /// <inheritdoc />
    public async Task UpdateAsync(int id, MenuItemUpsertDto upsertDto)
    {
        try
        {
            int updatedRecordCount = await _context.MenuItems
                .Where(mi => mi.Id == id)
                .ExecuteUpdateAsync(setters => setters
                    .SetProperty(mi => mi.Name, upsertDto.Name)
                    .SetProperty(mi => mi.DefaultNetPrice, upsertDto.DefaultNetPrice)
                    .SetProperty(mi => mi.DefaultVatPercentage, upsertDto.DefaultVatPercentage)
                    .SetProperty(mi => mi.Unit, upsertDto.Unit)
                    .SetProperty(mi => mi.ThumbnailUrl, upsertDto.ThumbnailUrl)
                    .SetProperty(
                        mi => mi.LastUpdatedDateTime,
                        DateTime.UtcNow.ToApplicationTime()));
            if (updatedRecordCount == 0)
            {
                throw new NotFoundException();
            }
        }
        catch (MySqlException exception)
        {
            bool isConvertible = TryConvertingDbException(
                exception,
                out OperationException convertedException);
            if (!isConvertible)
            {
                throw;
            }

            throw convertedException;
        }
    }

    /// <inheritdoc />
    public async Task DeleteAsync(int id)
    {
        int deletedRecordCount = 0;
        try
        {
            deletedRecordCount = await _context.MenuItems
                .Where(mi => mi.Id == id)
                .ExecuteDeleteAsync();
        }
        catch (MySqlException exception)
        {
            MySqlExceptionHandledResult handledResult = _exceptionHandler.Handle(exception);
            if (!handledResult.IsDeleteOrUpdateRestricted)
            {
                throw;
            }

            int updatedRecordCount = await _context.MenuItems
                .Where(mi => mi.Id == id)
                .ExecuteUpdateAsync(setters => setters
                    .SetProperty(mi => mi.IsDeleted, true));

            if (updatedRecordCount == 0)
            {
                throw new NotFoundException();
            }
        }

        if (deletedRecordCount == 0)
        {
            throw new NotFoundException();
        }
    }

    /// <summary>
    /// Tries converting the exception thrown by the database during the creating or updating
    /// operation into the exception that describes the error by convention.
    /// </summary>
    /// <param name="exception">
    /// The <see cref="MySqlException"/>  instance thrown by the database during the operation.
    /// </param>
    /// <param name="convertedException">
    /// The converted exception if the conversion is successful. Otherwise, <c>null</c>.
    /// </param>
    /// <returns>
    /// A <see cref="bool"/> value indicates whether the conversion is successful.
    /// </returns>
    private bool TryConvertingDbException(
        MySqlException exception,
        out OperationException convertedException)
    {
        convertedException = null;
        MySqlExceptionHandledResult handledResult = _exceptionHandler.Handle(exception);

        if (handledResult.IsUniqueConstraintViolated)
        {
            convertedException = OperationException.Duplicated(nameof(MenuItemUpsertDto.Name));
            return true;
        }

        if (handledResult.IsForeignKeyNotFound)
        {
            convertedException = OperationException.NotFound(
                new object[] { nameof(MenuItemUpsertDto.CategoryId) },
                DisplayNames.MenuCategory);
            return true;
        }

        return false;
    }
}